File: kolega_code/agent/eval/prelude.js
/**
 * Kolega Code eval-kernel prelude (JavaScript).
 *
 * Executed once in the kernel's persistent vm context at startup (sent by the
 * host as a silent init cell). Defines the model-facing in-kernel API:
 * display(), read()/write()/env(), the tool.<name>(args) bridge proxy,
 * listTools(), parallel(), npm_install(), setGlobal(), log()/phase().
 *
 * Style: async; helpers that hit the bridge return promises — use top-level
 * await. The tool proxy takes ONE trailing object literal, never positional.
 */

const __kcFs = require("node:fs");
const __kcPath = require("node:path");
const __kcCp = require("node:child_process");

async function __kolegaBridgeCall(name, args) {
  const base = process.env.KOLEGA_TOOL_BRIDGE_URL;
  const token = process.env.KOLEGA_TOOL_BRIDGE_TOKEN;
  const session = process.env.KOLEGA_TOOL_BRIDGE_SESSION;
  if (!base || !token || !session) throw new Error("tool bridge is unavailable in this kernel");
  const run = globalThis.__kolega_run_id__ || "";
  const response = await fetch(base.replace(/\/+$/, "") + "/v1/tool", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
    body: JSON.stringify({ session, run, name, args }),
  });
  let data;
  try {
    data = await response.json();
  } catch {
    throw new Error(`bridge call '${name}' failed: non-JSON response (HTTP ${response.status})`);
  }
  if (!data || data.ok !== true) {
    throw new Error((data && data.error) || `bridge call '${name}' failed`);
  }
  const value = data.value;
  if (
    value &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    Object.prototype.hasOwnProperty.call(value, "text") &&
    Object.keys(value).every((key) => key === "text" || key === "images")
  ) {
    return value.images && value.images.length ? value : value.text;
  }
  return value;
}

const tool = new Proxy(
  {},
  {
    get: (_target, name) => {
      if (typeof name !== "string" || name.startsWith("_")) return undefined;
      return (args) => __kolegaBridgeCall(name, args ?? {});
    },
  },
);

async function listTools() {
  return __kolegaBridgeCall("__list_tools__", {});
}

function display(value) {
  __kolega_display__(value);
}

function log(message) {
  __kolega_status__("log", { message: String(message) });
}

function phase(title) {
  __kolega_status__("phase", { title: String(title) });
}

function read(path, offset = 1, limit = undefined) {
  let data = __kcFs.readFileSync(String(path), "utf-8");
  if (offset > 1 || limit !== undefined) {
    const lines = data.split(/(?<=\n)/);
    const start = Math.max(0, offset - 1);
    const end = limit ? start + limit : lines.length;
    data = lines.slice(start, end).join("");
  }
  __kolega_status__("read", { path: String(path), chars: data.length });
  return data;
}

function write(path, content) {
  const target = __kcPath.resolve(String(path));
  __kcFs.mkdirSync(__kcPath.dirname(target), { recursive: true });
  __kcFs.writeFileSync(target, String(content), "utf-8");
  __kolega_status__("write", { path: target, chars: String(content).length });
  return target;
}

function env(key = undefined, value = undefined) {
  if (key === undefined) return { ...process.env };
  if (value !== undefined) {
    process.env[String(key)] = String(value);
    return value;
  }
  return process.env[String(key)];
}

function setGlobal(name, value) {
  globalThis[name] = value;
  return value;
}

function parallel(thunks) {
  return Promise.all(thunks.map((thunk) => thunk()));
}

function npm_install(...packages) {
  if (!packages.length) throw new Error("npm_install(...) needs at least one package name");
  const cmd = JSON.parse(process.env.KOLEGA_EVAL_NPM_INSTALL_CMD || "[]");
  if (!cmd.length) throw new Error("npm_install is unavailable in this kernel");
  const result = __kcCp.spawnSync(cmd[0], [...cmd.slice(1), ...packages.map(String)], {
    encoding: "utf-8",
  });
  if (result.error) throw result.error;
  if (result.status !== 0) {
    throw new Error(`npm_install failed (exit ${result.status}):\n${String(result.stderr).slice(-2000)}`);
  }
  __kolega_status__("npm_install", { packages });
  return String(result.stdout).slice(-2000);
}

File: kolega_code/agent/eval/prelude.py
"""Kolega Code eval-kernel prelude (Python).

Executed once in the kernel's persistent namespace at startup (sent by the host
as a silent init cell). Defines the model-facing in-kernel API: display(),
read()/write()/env(), the tool.<name>(args) bridge proxy, list_tools(),
parallel(), pip_install(), python_info(), log()/phase().

Stdlib only. Runner provides __kolega_display__, __kolega_status__, and the
__kolega_run_id__ global (updated per cell).
"""

from __future__ import annotations

import json as _json
import os as _os
import subprocess as _subprocess
import urllib.error as _urlerror
import urllib.request as _urlrequest
from concurrent.futures import ThreadPoolExecutor as _ThreadPoolExecutor
from pathlib import Path as _Path


def display(value):
    """Render a value as a rich output (images, HTML, JSON) in the tool result."""
    # __kolega_display__ is injected into the kernel namespace by the runner.
    globals()["__kolega_display__"](value)


def _emit_status(op, **data):
    # __kolega_status__ is injected into the kernel namespace by the runner.
    globals()["__kolega_status__"](op, **data)


def log(message):
    """Add a note to this cell's status output."""
    _emit_status("log", message=str(message))


def phase(title):
    """Mark a named phase of this cell in its status output."""
    _emit_status("phase", title=str(title))


def env(key=None, value=None):
    """Get or set kernel environment variables."""
    if key is None:
        return dict(sorted(_os.environ.items()))
    if value is not None:
        _os.environ[str(key)] = str(value)
        return value
    return _os.environ.get(str(key))


def read(path, offset=1, limit=None):
    """Read a file's contents. offset/limit are 1-indexed lines."""
    p = _Path(path)
    data = p.read_text(encoding="utf-8")
    if offset > 1 or limit is not None:
        lines = data.splitlines(keepends=True)
        start = max(0, int(offset) - 1)
        end = start + int(limit) if limit else len(lines)
        data = "".join(lines[start:end])
    _emit_status("read", path=str(p), chars=len(data))
    return data


def write(path, content):
    """Write text to a file, creating parent directories. Returns the path."""
    p = _Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(str(content), encoding="utf-8")
    _emit_status("write", path=str(p), chars=len(str(content)))
    return p


def _bridge_config():
    base = _os.environ.get("KOLEGA_TOOL_BRIDGE_URL")
    token = _os.environ.get("KOLEGA_TOOL_BRIDGE_TOKEN")
    session = _os.environ.get("KOLEGA_TOOL_BRIDGE_SESSION")
    if not base or not token or not session:
        raise RuntimeError("tool bridge is unavailable in this kernel")
    return base.rstrip("/"), token, session


def _bridge_call(name, args):
    """POST one request to the host tool bridge and return its value."""
    base, token, session = _bridge_config()
    run_id = globals().get("__kolega_run_id__") or ""
    payload = _json.dumps({"session": session, "run": run_id, "name": name, "args": args}).encode("utf-8")
    request = _urlrequest.Request(
        f"{base}/v1/tool",
        data=payload,
        method="POST",
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {token}"},
    )
    try:
        with _urlrequest.urlopen(request) as response:
            body = response.read()
    except _urlerror.HTTPError as exc:
        body = exc.read()
    except _urlerror.URLError as exc:
        raise RuntimeError(f"bridge call {name!r} failed: {exc}") from None
    try:
        data = _json.loads(body)
    except _json.JSONDecodeError:
        raise RuntimeError(f"bridge call {name!r}: non-JSON response: {body[:200]!r}") from None
    if not isinstance(data, dict) or not data.get("ok"):
        message = data.get("error") if isinstance(data, dict) else None
        raise RuntimeError(message or f"bridge call {name!r} failed")
    return data.get("value")


def _unwrap_tool_value(value):
    """Tool results are plain text unless images are attached."""
    if isinstance(value, dict) and "text" in value and set(value) <= {"text", "images"}:
        if value.get("images"):
            return value
        return value["text"]
    return value


class _ToolCallable:
    """Invokes one host-side tool via the loopback bridge."""

    __slots__ = ("_name",)

    def __init__(self, name):
        self._name = name

    def __repr__(self):
        return f"<tool.{self._name}>"

    def __call__(self, args=None, /, **kwargs):
        if args is None:
            merged = {}
        elif isinstance(args, dict):
            merged = dict(args)
        else:
            raise TypeError(f"tool.{self._name}(...) expects a dict of arguments (got {type(args).__name__})")
        merged.update(kwargs)
        return _unwrap_tool_value(_bridge_call(self._name, merged))


class _ToolProxy:
    """``tool.<name>(args)`` proxy over the host's tool registry."""

    __slots__ = ()

    def __getattr__(self, name):
        if name.startswith("_"):
            raise AttributeError(name)
        return _ToolCallable(name)

    def __getitem__(self, name):
        return _ToolCallable(name)

    def __dir__(self):
        try:
            return sorted(entry.get("name", "") for entry in list_tools() or [])
        except Exception:
            return []

    def __repr__(self):
        return "<tool proxy: call list_tools() to discover available tools>"


tool = _ToolProxy()


def list_tools():
    """List the tools available through `tool.<name>` as [{name, summary}]."""
    return _bridge_call("__list_tools__", {})


def parallel(thunks, max_workers=None):
    """Run callables concurrently (threads) and return their results in order."""
    thunks = list(thunks)
    with _ThreadPoolExecutor(max_workers=max_workers or max(1, len(thunks))) as pool:
        futures = [pool.submit(thunk) for thunk in thunks]
        return [future.result() for future in futures]


def pip_install(*packages):
    """Install packages into the kernel's dedicated Python environment."""
    if not packages:
        raise ValueError("pip_install(...) needs at least one package name")
    cmd = _json.loads(_os.environ.get("KOLEGA_EVAL_PIP_INSTALL_CMD") or "[]")
    if not cmd:
        raise RuntimeError(
            "pip_install is unavailable: the kernel is not running in a managed environment "
            "(custom interpreter or provisioning was skipped). Install packages into that "
            "interpreter yourself, e.g. via tool.exec_command."
        )
    proc = _subprocess.run(list(cmd) + [str(p) for p in packages], capture_output=True, text=True, check=False)
    if proc.returncode != 0:
        raise RuntimeError(f"pip_install failed (exit {proc.returncode}):\n{proc.stderr[-2000:]}")
    _emit_status("pip_install", packages=list(packages))
    return (proc.stdout or "")[-2000:]


def python_info():
    """Return interpreter version, env path, and bundled packages for this kernel."""
    import sys as _sys

    return {
        "version": _sys.version.split()[0],
        "executable": _sys.executable,
        "env_path": _os.environ.get("KOLEGA_EVAL_ENV_PATH") or None,
        "bundle": [p for p in (_os.environ.get("KOLEGA_EVAL_BUNDLE") or "").split(",") if p],
    }
